### ISSUE TITLE
Requirements could not be resolved, local dependencies

### ISSUE THREAD

**Issue by jannes-io** [composer/composer#12064]
My `composer.json`:

Test project (local development sandbox):
```json
{
    "require": {
        "vendor/package-1": "*",
        "vendor/package-2": "*"
    },
    "repositories": [
        {
            "type": "path",
            "url": "../package-1"
        },
        {
            "type": "path",
            "url": "../package-2"
        }
    ]
}
```

Package-2 (published to packagist):
```json
{
    "require": {
        "vendor/package-1": "^1.2.3",
    }
}
```

When I run this command: 

```
composer update vendor/package-1
```

I get the following output: 

```
Your requirements could not be resolved to an installable set of packages.

  Problem 1
    - Root composer.json requires vendor/package-2 * -> satisfiable by vendor/package-2[dev-master].
    - vendor/package-2 dev-master requires vendor/package-1 ^1.2.3 -> found vendor/package-1[dev-master] but it does not match the constraint.

```

And I expected this to happen:

Ignore version constraints when installing from a local repository and no version is set in any composer.json (using git tags to version).

**Context**:
I have a plugin (package-2 in this example) that has a version constraint on my framework (package-1). Currently it's impossible to add dependencies to package-1 and run `composer update` in my development project without using some workaround.

Known workarounds:
- Temporarily set a version in composer.json of package-1.
	- If forgotten about and pushed to remote, packagist will most likely reject the version.
- Remove the version constraint of package-1 in package-2's composer.json.
	- if forgotten about package-1 might get updated to a version that is not supported by package-2.

Tl;dr asking for a DX improvement to skip constraint checking when using local unversioned packages so we can define our constraints in dependencies when published to packagist, but work on dev-master locally.

**Comment by Seldaek** [comment]
Have you tried to set an extra.branch-alias in package-1 so that dev-master is seen as  1.2.x-dev or 1.x-dev? That should solve it I believe.

**Comment by jannes-io** [comment]
@Seldaek That does work for 1 version.

**Comment by Seldaek** [comment]
Sorry not sure what you are implying. It works but not fully? If not then what/why?

**Comment by jannes-io** [comment]
> Sorry not sure what you are implying. It works but not fully? If not then what/why?

What I meant is it will work for 1 version, but when that package gets a major/minor update, it'll stop working. Been using branch alias since your suggestion and it hasn't caused any further issues, just need to remember to update the branch aliases when I release a new minor/major.
